pkg/urcrypt/urcrypt/pedersen.c

```diff
@@ -0,0 +1,356 @@
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  prj_pt sum;
+  ret = prj_pt_add(&sum, &(cxt->cp->p0), &p1_alow);
+  if ( 0 != ret ) {
+    return ret;
+  }
+  ret = prj_pt_add(&sum, &sum, &p2_ahig);
+  if ( 0 != ret ) {
+    return ret;
+  }
+  ret = prj_pt_add(&sum, &sum, &p3_blow);
+  if ( 0 != ret ) {
+    return ret;
+  }
+  ret = prj_pt_add(&sum, &sum, &p4_bhig);
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  prj_pt aff;
+  ret = prj_pt_unique(&aff, &sum);
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  ret = fp_export_to_buf(out, 32, &(aff.X));
+  urcrypt__reverse(32, out);
+  return ret;
+}
+
+static int
+do_fold_hash(urcrypt_pedersen_context *cxt, uint8_t *dat_x, size_t len_x, uint8_t *out) {
+  int ret = 0;
+  int pos = 0, ext_pos = 0, ext_len = (len_x / 32);
+  uint8_t *hed = NULL, *acc = NULL, *ext = NULL;
+  while ( pos < len_x ) {
+    hed = (dat_x + pos);
+    if ( (len_x - pos) >= 32 && (hed[31] & 0xf0) != 0) {
+        if ( ext == NULL ) {
+          ext = calloc(1, ext_len);
+        }
+        ext[ext_pos] = (hed[31] & 0xf0) >> 4;
+        ext_pos++;
+        hed[31] = hed[31] & 0x0f;
+      }
+    if ( 0 == pos) {
+      acc = hed;
+    }
+    else {
+      ret = do_hash(cxt, acc, 32, hed, MIN(len_x - pos, 32), out);
+      if ( ret != 0 ) {
+        if ( ext != NULL ) {
+          free(ext);
+        }
+        return ret;
+      }
+      acc = out;
+    }
+    pos += 32;
+  }
+  if ( ext != NULL ) {
+    ret = do_hash(cxt, acc, 32, ext, ext_len, out);
+    free(ext);
+  }
+  return ret;
+}
+
+int
+urcrypt_pedersen(urcrypt_pedersen_context *cxt, uint8_t *a, size_t a_len, uint8_t *b, size_t b_len, uint8_t out_y[32])
+{
+  int ret = 0;
+  uint8_t a_vat[32], b_vat[32], *a_dat, *b_dat;
+  size_t a_ven, b_ven;
+  if ( ( a_len < 32 ) ||
+       ((32 == a_len) && ((a[31] & 0xf0) == 0)) )
+  {
+      // a <= 252 bits so use directly
+      a_ven = a_len;
+      a_dat = a;
+  }
+  else {
+    // a > 252 bits so break up into chunks and fold hash across
+    a_ven = 32;
+    ret = do_fold_hash(cxt, a, a_len, a_vat);
+    if ( 0 != ret ) {
+      return ret;
+    }
+    a_dat = a_vat;
+  }
+
+  if ( (b_len < 32) ||
+       ((32 == b_len) && ((b[31] & 0xf0) == 0)) )
+  {
+      // b <= 252 bits
+      b_ven = b_len;
+      b_dat = b;
+  }
+  else {
+    // b > 252 bits so break up into chunks and fold hash across
+    b_ven = 32;
+    ret = do_fold_hash(cxt, b, b_len, b_vat);
+    if ( 0 != ret ) {
+      return ret;
+    }
+    b_dat = b_vat;
+  }
+
+  return do_hash(cxt, a_dat, a_ven, b_dat, b_ven, out_y);
+}
+
+int
+urcrypt_pedersen_init(urcrypt_pedersen_context *context)
+{
+  int ret;
+  context->cp = malloc(sizeof(constant_points));
+  context->curve_params = malloc(sizeof(ec_params));
+  ret = get_curve_params(context->curve_params);
+  if ( 0 != ret ) {
+    return -1;
+  }
+  return get_constant_points(context->curve_params, context->cp);
+}
+
+void
+urcrypt_pedersen_destroy(urcrypt_pedersen_context *context)
+{
```

Comment:
probably you don't need destroy function at all if the context isn't malloc'd, but you also didn't free the context here.